src/librustc/ty/instance.rs

```diff
@@ -219,6 +224,13 @@ fn resolve_associated_item<'a, 'tcx>(
         traits::VtableImpl(impl_data) => {
             let (def_id, substs) = traits::find_associated_item(
                 tcx, trait_item, rcvr_substs, &impl_data);
+
+            check_unimplemented_trait_item(tcx,
```

Comment:
This doesn't seem like the right place to be doing this check. This function runs after type-checking has completed, so this error should be reported earlier. Under what scenario did this code execute before?

Reply:
At first I thought that this error should be raised in typeck but I didn't found in there where an impl method call is evaluated against a set of possible impls.
I mean, given this code 
```
 trait Foo {
     fn foo_one(&self) -> &'static str;
     fn foo_two(&self) -> &'static str;
 }
 
 struct MyStruct;
 
 default impl<T> Foo for T {
     fn foo_one(&self) -> &'static str {
         "generic"
     }
 }
 
 default impl<T: Clone> Foo for T {
     fn foo_one(&self) -> &'static str {
         "generic Clone"
     }
 }
```

If i call `MyStruct.foo_one()` the compiler select the method contained in the first impl. 
The error should be raised in case this impl doesn't contain an imlementation of `foo_one`.
Since I didn't found the selection process during typeck I've choosen to implement it after it occurs inside the `resolve_associated_item` function

Do you think I should move all inside typeck?

Reply:
right, so what should happen in this case is that you get an error, because `MyStruct` does not implement `Foo`. This is sort of related to the prior discussion -- having a `default impl` just means that you have some defaults that are applied *to other impls*. 

What I would expect is that if the user added a `impl Foo for MyStruct`, then we would get an error there saying that no version of `foo_two` can be found.

Reply:
Ok, Got it. So there is no need to check for an unimplemented trait item. I'll remove it and include some tests on the use cases you provided.